granulate_utils/logging/__init__.py

```diff
@@ -0,0 +1,138 @@
+        self.buffer: List[dict] = []
+        self.server_address = server_address
+        self.session = requests.Session()
+        self.stop_event = threading.Event()
+        self.time_fn = time.time
+        self.last_flush_time = 0.0
+        self.flush_thread = threading.Thread(target=self._flush_loop, daemon=True, name="Logs flusher")
+        self.setup()
+        self.flush_thread.start()
+
+    def setup(self) -> None:
+        """Called to perform setup before the flush thread is started."""
+        pass
+
+    def emit(self, record: LogRecord) -> None:
+        # Called while lock is acquired
+        item = self.dictify_record(record)
+        self.buffer.append(item)
+        self.next_serial_no += 1
+        self.check_overflow()
+
+    def check_overflow(self) -> None:
+        if len(self.buffer) >= self.capacity:
+            drop_count = int(self.overflow_drop_factor * self.capacity)
+            del self.buffer[:drop_count]
+
+    def dictify_record(self, record: LogRecord) -> Dict[str, Any]:
+        # `format` is required to fill in some record members such as `message` and `asctime`.
+        self.format(record)
+        formatted_timestamp = datetime.utcfromtimestamp(record.created).isoformat()
+        return {
+            "serial_no": self.next_serial_no,
+            "message": record.message,
+            "level": record.levelname,
+            "timestamp": formatted_timestamp,
+            "logger_name": record.name,
+        }
+
+    def _flush_loop(self) -> None:
+        self.last_flush_time = self.time_fn()
+        while not self.stop_event.is_set():
+            if self.should_flush():
+                self._flush()
+            self.stop_event.wait(1.0)
+
+    def should_flush(self) -> bool:
+        delta_time = self.time_fn() - self.last_flush_time
+        return (len(self.buffer) >= self.flush_threshold * self.capacity) or (delta_time >= self.flush_interval)
+
+    def _flush(self) -> None:
+        self.last_flush_time = self.time_fn()
+        try:
+            response = self._send_logs()
+            response.raise_for_status()
+            last_ingested_serial_no = response.json()["last_ingested_serial_no"]
+        except Exception:
+            logger.error("Error posting to server")
+        else:
+            self.remove_up_to_serial_no(last_ingested_serial_no + 1)
+
+    @backoff.on_exception(backoff.expo, exception=RequestException, max_tries=REQUEST_MAX_TRIES)
+    def _send_logs(self) -> requests.Response:
+        # Upon every retry we will remake the batch, in case we are able to batch more messages together.
+        data = {
+            "id": uuid.uuid4().hex,
+            "metadata": self.get_metadata(),
+            "logs": self.make_batch(),
+        }
+        return self.session.post(f"{self.scheme}://{self.server_address}/", json=data, timeout=self.request_timeout)
+
+    def get_metadata(self) -> dict:
+        """Called to get metadata per batch."""
+        return {}
```

Comment:
Need to implement basic ones and be able to augment with additional metadata from the ctor (so that agents will be able to add more custom metadata)